core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -519,20 +526,22 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
       val newMessageBuffer = newBuffer.slice()
       newMessageBuffer.limit(newMessageSize)
       message.convertToBuffer(toMagicValue, newMessageBuffer, now, timestampType)
-
+      if (toMagicValue > Message.MagicValue_V0)
+        maxTimestamp = math.max(newMessageBuffer.getLong(Message.TimestampOffset), maxTimestamp)
       newMessagePosition += MessageSet.LogOverhead + newMessageSize
     }
     newBuffer.rewind()
-    new ByteBufferMessageSet(newBuffer)
+    (new ByteBufferMessageSet(newBuffer), maxTimestamp)
```

Comment:
It would be useful to add a comment explaining the return type before the method.